File: leetcode/19.remove-nth-node-from-end-of-list.go
/*
 * @lc app=leetcode id=19 lang=golang
 *
 * [19] Remove Nth Node From End of List
 */

type ListNode struct {
	Val  int
	Next *ListNode
}

// @lc code=start

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	length := 0

	var dfs func(node *ListNode, cur int)
	dfs = func(node *ListNode, cur int) {
		if node == nil {
			length = cur
			return
		}

		dfs(node.Next, cur+1)

		if length-n-1 == cur {
			if node.Next != nil {
				node.Next = node.Next.Next
			} else {
				node.Next = nil
			}
		}
	}
	dfs(head, 0)

	if length == 1 {
		return nil
	} else if n == length {
		return head.Next
	}

	return head
}

// @lc code=end

File: leetcode/19.remove-nth-node-from-end-of-list.cpp
/*
 * @lc app=leetcode id=19 lang=cpp
 *
 * [19] Remove Nth Node From End of List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int len = 0;
        function<int(ListNode*)> dsf = [&](ListNode* node) -> int {
            if (node == NULL) {
                return 1;
            } else {
                len++;
                int curNum = dsf(node->next);
                
                if (curNum == n) return -1;
                else if (curNum == -1) {
                    node->next = node->next->next;
                    return -2;
                }
                else if (curNum == -2) return -2;

                return curNum + 1;
            }

        };
        dsf(head);

        if (len-n > 0) return head;
        return head->next;
    }
};
// @lc code=end

